driver-async/src/main/com/mongodb/async/client/gridfs/GridFSBucketImpl.java

```diff
@@ -0,0 +1,401 @@
+        downloadToAsyncOutputStream(openDownloadStream(id), destination, errorHandlingCallback(callback));
+    }
+
+    @Override
+    public GridFSDownloadStream openDownloadStream(final BsonValue id) {
+        return new GridFSDownloadStreamImpl(find(new Document("_id", id)), chunksCollection);
+    }
+
+    @Override
+    public void downloadToStream(final BsonValue id, final AsyncOutputStream destination, final SingleResultCallback<Long> callback) {
+        downloadToAsyncOutputStream(openDownloadStream(id), destination, errorHandlingCallback(callback));
+    }
+
+    @Override
+    public GridFSDownloadStream openDownloadStreamByName(final String filename) {
+        return openDownloadStreamByName(filename, new GridFSDownloadByNameOptions());
+    }
+
+    @Override
+    public GridFSDownloadStream openDownloadStreamByName(final String filename, final GridFSDownloadByNameOptions options) {
+        return new GridFSDownloadStreamImpl(findFileByName(filename, options), chunksCollection);
+    }
+
+    @Override
+    public void downloadToStreamByName(final String filename, final AsyncOutputStream destination,
+                                       final SingleResultCallback<Long> callback) {
+        downloadToStreamByName(filename, destination, new GridFSDownloadByNameOptions(), callback);
+    }
+
+    @Override
+    public void downloadToStreamByName(final String filename, final AsyncOutputStream destination,
+                                       final GridFSDownloadByNameOptions options, final SingleResultCallback<Long> callback) {
+        downloadToAsyncOutputStream(openDownloadStreamByName(filename, options), destination, errorHandlingCallback(callback));
+    }
+
+    @Override
+    public GridFSFindIterable find() {
+        return new GridFSFindIterableImpl(filesCollection.find());
+    }
+
+    @Override
+    public GridFSFindIterable find(final Bson filter) {
+        return new GridFSFindIterableImpl(filesCollection.find(filter));
+    }
+
+    @Override
+    public void delete(final ObjectId id, final SingleResultCallback<Void> callback) {
+        final SingleResultCallback<Void> errorHandlingCallback = errorHandlingCallback(callback);
+        filesCollection.deleteOne(new BsonDocument("_id", new BsonObjectId(id)), new SingleResultCallback<DeleteResult>() {
+            @Override
+            public void onResult(final DeleteResult filesResult, final Throwable t) {
+                if (t != null) {
+                    errorHandlingCallback.onResult(null, t);
+                } else {
+                    chunksCollection.deleteMany(new BsonDocument("files_id", new BsonObjectId(id)),
+                            new SingleResultCallback<DeleteResult>() {
+                                @Override
+                                public void onResult(final DeleteResult chunksResult, final Throwable t) {
+                                    if (filesResult.wasAcknowledged() && filesResult.getDeletedCount() == 0) {
```

Comment:
This works slightly differently from the sync version of this method, which will throws an exception from deleting from the chunks collection before checking deleted count on the files collection.